$OpenBSD: patch-libavcodec_aacenc_is_c,v 1.1 2015/09/12 09:44:54 ajacoutot Exp $

aacenc_is: ignore PNS band types

aacenc_is: take absolute coefficient value upon energy calculations

--- libavcodec/aacenc_is.c.orig	Sat Sep 12 02:07:50 2015
+++ libavcodec/aacenc_is.c	Sat Sep 12 02:08:20 2015
@@ -107,10 +107,15 @@ void ff_aac_search_for_is(AACEncContext *s, AVCodecCon
                 cpe->ch[1].band_type[w*16+g] != NOISE_BT && !cpe->ch[1].zeroes[w*16+g]) {
                 float ener0 = 0.0f, ener1 = 0.0f, ener01 = 0.0f;
                 struct AACISError ph_err1, ph_err2, *erf;
+                if (sce0->band_type[w*16+g] == NOISE_BT ||
+                    sce1->band_type[w*16+g] == NOISE_BT) {
+                    start += sce0->ics.swb_sizes[g];
+                    continue;
+                }
                 for (w2 = 0; w2 < sce0->ics.group_len[w]; w2++) {
                     for (i = 0; i < sce0->ics.swb_sizes[g]; i++) {
-                        float coef0 = sce0->pcoeffs[start+(w+w2)*128+i];
-                        float coef1 = sce1->pcoeffs[start+(w+w2)*128+i];
+                        float coef0 = fabsf(sce0->pcoeffs[start+(w+w2)*128+i]);
+                        float coef1 = fabsf(sce1->pcoeffs[start+(w+w2)*128+i]);
                         ener0  += coef0*coef0;
                         ener1  += coef1*coef1;
                         ener01 += (coef0 + coef1)*(coef0 + coef1);
